crates/kubelet/src/config.rs

```diff
@@ -85,77 +122,240 @@ impl Config {
+    #[cfg_attr(feature = "docs", doc(cfg(feature = "cli")))]
+    fn new_from_file_and_flags_impl(version: &str, config_file_path: PathBuf) -> Self {
+        // TODO: reduce duplication
+        let app = Opts::clap().version(version);
+        let opts = Opts::from_clap(&app.get_matches());
+        let cli_builder = ConfigBuilder::from_opts(opts);
+
+        let config_file_builder = ConfigBuilder::from_config_file(config_file_path);
+
+        let builder = config_file_builder.unwrap().with_override(cli_builder); // if the config file is actually malformed then we should halt even if there are CLI values
+        Config::new_from_builder(builder)
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self::default_config(
+            &"127.0.0.1"
+                .parse()
+                .expect("Could not parse hardcoded address"),
+        )
+        .expect("Could not create default config")
+    }
+}
+
+fn ok_result_of<T>(value: Option<T>) -> Option<anyhow::Result<T>> {
+    value.map(Ok)
+}
 
-        let node_labels = opts
+impl ConfigBuilder {
+    fn from_opts(opts: Opts) -> Self {
+        let node_labels: Vec<(String, String)> = opts
             .node_labels
             .iter()
             .filter_map(|i| split_one_label(i))
             .collect();
 
-        let port = opts.port;
+        ConfigBuilder {
+            node_ip: ok_result_of(opts.node_ip),
+            node_name: opts.node_name,
+            node_labels: if node_labels.is_empty() {
+                None
+            } else {
+                Some(HashMap::from_iter(node_labels))
+            },
+            hostname: opts.hostname,
+            data_dir: opts.data_dir,
+            max_pods: ok_result_of(opts.max_pods),
+            server_addr: ok_result_of(opts.addr),
+            server_port: ok_result_of(opts.port),
+            server_tls_cert_file: opts.tls_cert_file,
+            server_tls_private_key_file: opts.tls_private_key_file,
+        }
+    }
+
+    fn from_config_file(config_file_path: PathBuf) -> anyhow::Result<ConfigBuilder> {
+        if !config_file_path.exists() {
+            return Ok(ConfigBuilder::default());
+        }
+        let config_source = config_file::File::from(config_file_path);
+        ConfigBuilder::from_config_source(config_source)
+    }
 
-        let data_dir = opts
-            .data_dir
-            .unwrap_or_else(|| default_data_dir().expect("unable to get default directory"));
+    fn from_config_source<T>(source: T) -> anyhow::Result<ConfigBuilder>
+    where
+        T: 'static,
+        T: config_file::Source + Send + Sync,
+    {
+        let mut settings = config_file::Config::default();
+        match settings.merge(source) {
+            Ok(s) => Ok(ConfigBuilder::from_config_settings(s.clone())),
+            Err(e) => Err(anyhow::Error::new(e)),
+        }
+    }
+
+    fn from_config_settings(settings: config_file::Config) -> ConfigBuilder {
+        let port = settings
```

Comment:
Did you by chance look at using [`try_into`](https://docs.rs/config/0.10.1/config/struct.Config.html#method.try_into) and deserializing it to a concrete data structure instead of having to do all these lookups?

Reply:
I hadn't - thanks!  This seems like it should be the way to go, but `try_into` requires the target type to implement `Deserialize`.  And I can't derive `Deserialize` on ConfigBuilder because serde doesn't know how to handle the `Option<Result<T>>` traits, and if I go back to straight `Option<T>` then I will get a different behaviour, namely that _any_ invalid value in the config file renders the entire config file unusable, even if that value is overridden on the command line.  Maybe that's an acceptable trade-off (maybe it's even desirable).

Of course I could implement `Deserialize` manually but then I'd be back to doing all the lookups so there's probably not much value in that!

Reply:
Or maybe there are attributes I can apply to the `<Option<Result<T>>` fields that would get the desired behaviour...

Reply:
There are some attributes I think. An option result is a bit weird but custom deserialization isn’t too bad if I remember correctly from doing it myself

Reply:
BY THE HOARY HOSTS OF HOGGOTH @thomastaylor312 YOU ARE RIGHT.  Works out significantly nicer this way.